cpp/src/fil/treelite_import.cu

```diff
@@ -349,69 +366,55 @@ int tree2fil(std::vector<fil_node_t>& nodes,
              std::size_t* leaf_counter,
              cat_sets_owner* cat_sets)
 {
-  typedef std::pair<int, int> pair_t;
-  std::stack<pair_t> stack;
   // needed if the node is sparse, to place within memory for the FIL tree
-  int sparse_index = root + 1;
-  stack.push(pair_t(tree_root(tree), 0));
-  while (!stack.empty()) {
-    const pair_t& top = stack.top();
-    int node_id       = top.first;
-    int cur           = top.second;
-    stack.pop();
-
-    while (!tree.IsLeaf(node_id)) {
+  int sparse_index = 1;
+  walk_tree(
+    tree,
+    int(0),  // descent accumulator (FIL node ID) initial value
+    [&](int node_id, int fil_node_id) {
       // reserve space for child nodes
       // left is the offset of the left child node relative to the tree root
       // in the array of all nodes of the FIL sparse forest
-      int left = node_traits<fil_node_t>::IS_DENSE ? 2 * cur + 1 : sparse_index - root;
+      int left = node_traits<fil_node_t>::IS_DENSE ? 2 * fil_node_id + 1 : sparse_index;
       sparse_index += 2;
       conversion_state<fil_node_t> cs = tl2fil_inner_node<fil_node_t>(
         left, tree, node_id, cat_sets, &cat_sets->bit_pool_offsets[tree_idx]);
-      nodes[root + cur] = cs.node;
-      // push child nodes into the stack
-      stack.push(pair_t(cs.tl_right, left + 1));
-      // stack.push(pair_t(tl_left, left));
-      node_id = cs.tl_left;
-      cur     = left;
-    }
-
-    // leaf node
-    nodes[root + cur] = fil_node_t({}, {}, 0, false, true, false, 0);
-    tl2fil_leaf_payload(
-      &nodes[root + cur], root + cur, tree, node_id, forest_params, vector_leaf, leaf_counter);
-  }
-
+      nodes[root + fil_node_id] = cs.node;
+
+      return std::pair<int, int>(left + cs.swap_child_nodes, left + !cs.swap_child_nodes);
```

Comment:
`left + (cs.swap_child_nodes ? 1 : 0)`, and similar with the other expression.

Reply:
is the new version simple?